designs/error-reporting-design.md

```diff
@@ -0,0 +1,341 @@
+<!-- # TODOs:
+- Review everything in "Examples of Design"
+- Review "Changes to Data Structure".
+- Review "Functions to add"
+- Review "Future Work" -->
+
+# Error Reporting
+
+## Design
+### What we aim to achieve by next week
+1. More information in all error messages
+    1. [Report Line Number](#report-line-number)
+    2. [Report Column Number](#report-column-number)
+    3. [Report Source Code](#report-source-code)
+    4. [Pretty Source Code](#pretty-source-code)
+
+2. Better type error
+    1. [Assignment with different types](#Assignment-with-different-types)
+    2. [Override \_\_init\_\_ signature](#override-\_\_init\_\_-signature)
+    3. [Binary operator type hint](#binary-operator-type-hint)
+    4. [Condition expression type hint](#condition-expression-type-hint)
+    5. [Type check print/len argument](#type-check-print/len-argument)
+
+3. Better parse error
+    1. [Using keywords as variable names](#using-keywords-as-variable)
+
+4. Better Run-time error
+    1. [Divide by zero](#divide-by-zero)
+
+### Future Work
+1. [Stack trace](#stack-trace)
+
+2. Better type error
+    1. [Function return check](#function-return-check)
+    2. [Branch return hint](#branch-return-hint)
+
+2. Better Run-time error
+    1. [Assert not none](#assert-not-none)
+    2. [Out of memory](#out-of-memory)
+
+
+
+## Changes to Data Structures
+
+The only changes we would add to the AST is the following:
+
+- We're extending the annotation type `A` to `Annotation` with information of start and end location (line and column numbers) as well as the full line of source code that contains the given component.
+- Such information except for the type information will be added during parsing, and type information will be added after type checking.
+
+Here's an overview of the changes in code:
+
+```ts
+Annotation = {
+    type?: Type,
+    fromLoc: Loc,
+    toLoc: Loc,
+    fullSrc: string
+}
```

Comment:
This will involve changing lots of code that sets the `a` field, right? (I don't know that I have a great way for you to get around that, just making sure).

Reply:
We can get around that by keeping the current `a` field as is and adding a new field to every AST node that stores the location info we want. 
But we thought that maybe other groups also want to annotate the AST with relevant information, and we might as well put everything in one place. It makes changing the annotation down the line easier. 
The decision depends on if we want to rewrite the previous type checking codes.